src/JustEat.StatsD/SpanStatsDMessageFormatter.cs

```diff
@@ -0,0 +1,266 @@
+using System;
+using System.Buffers;
+using System.Buffers.Text;
+using System.Globalization;
+using System.Runtime.CompilerServices;
+using System.Text;
+
+namespace JustEat.StatsD
+{
+    public static class FixedBufferExtensions
+    {
+        public static string Show(in this FixedBuffer src) => Encoding.UTF8.GetString(src.Written());
+
+        public static Span<byte> Written(in this FixedBuffer src) => src.Buffer.Slice(0, src.Position);
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static ref FixedBuffer Add(ref this FixedBuffer src, byte[] str)
+        {
+            str.CopyTo(src.Buffer.Slice(src.Position));
+            src.Position += str.Length;
+            return ref src;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static ref FixedBuffer Add(ref this FixedBuffer src, byte str)
+        {
+            src.Buffer[src.Position] = str;
+            src.Position += 1;
+            return ref src;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static ref FixedBuffer Add(ref this FixedBuffer src, string str)
+        {
+            var written = Encoding.UTF8.GetBytes(str, src.Buffer.Slice(src.Position));
+            src.Position += written;
+            return ref src;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static ref FixedBuffer Add(ref this FixedBuffer src, double value)
+        {
+            Span<char> dest = stackalloc char[32];
+            value.TryFormat(dest, out int written, "", CultureInfo.InvariantCulture);
+
+            
+
+            ref var pos = ref src.Position;
+            
+            for (var i = 0; i < written; i++)
+            {
+                src.Buffer[pos++] = (byte) dest[i];
+            }
+
+            return ref src;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static ref FixedBuffer Add(ref this FixedBuffer src, long value)
+        {
+            if (!Utf8Formatter.TryFormat(value, src.Buffer.Slice(src.Position), out int written, new StandardFormat('d')))
+            {
+                throw new OutOfMemoryException();
+            }
+
+            src.Position += written;
+            return ref src;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static void Clear(ref this FixedBuffer src)
+        {
+            src.Position = 0;
+        }
+    }
+
+    public ref struct FixedBuffer
+    {
+        public Span<byte> Buffer;
+        public int Position;
+
+        public FixedBuffer(Span<byte> buffer) : this()
+        {
+            Buffer = buffer;
+            Position = 0;
+        }
+    }
+
+    public class SpanStatsDMessageFormatter
+    {
+        private const double DefaultSampleRate = 1.0;
+
+        [ThreadStatic]
```

Comment:
Why would we need a `Random` per thread?

Reply:
Concurrent access to `Random` is unsafe and can gives weird results, I've seen it return lot's of `0`s under high load.

Reply:
Fair enough - I've never heard of that problem before.